### ISSUE TITLE
broken swap with dracut  / mkinitcpio + Luks swap 

### ISSUE THREAD

**Issue by abucodonosor** [calamares/calamares#730]
Hello,

when using Luks  entry for bootloader config and fstab/crypttab are wrong.

calamares would generate this 

grub:
... resume=UUID=the_swap_uuid rd.luks.uuid=the_luks_uuid

crypttab:
luks-the_luks_uuid        key          swap

fstab:
UUID=the_swap_uuid

with LUKS the resume= cmdline need point to the device with luks-swap eg: /dev/mapper/luks-xxxx

crypttab need swap,plain,luks as option so systemd or other generators can open it

fstab want swap to be /dev/mapper/luks-xxxx so systemd or whatever init manage swap can mount it

I would suggest to not mix all these but on LUKS use root = && resume=  /dev/mapper  and same for fstab

Regards

**Comment by demmm** [comment]
Possible related to https://github.com/calamares/calamares/issues/606

**Comment by kkofler** [comment]
This is really a `dracut` bug. `resume=UUID=…` is supposed to work and to be handled by the initramfs. `dracut` fails to handle `resume=…`, so the kernel sees it, and the kernel supports only direct /dev nodes.

**Comment by abucodonosor** [comment]
@kkofler 

resume=UUID= does work here for !LUKS 'as' is.. the problem is not dracut but the logic how cala handles
luks installs.

on luks && dracut cala does 
swap_uuid = ""
swap_outer_uuid = "" <-- for dracut

check swap && if found set resume=UUID=real-device-swap-UUID <-- which is not same as the LUKS swap UUID ofc

check LUKS and if found && dracut found set rd.luks.uuid=the-luks-uuid < all fine

missing check on if swap && LUKS in bootloader , fstab , grubcfg modules  but then it go an writes the following filese

fstab
UUID=root-one
UUID=broke-swap-uuid

kernel cmd

.... resume=UUID=broken-one rd.luks.uiid=right.one

now you boot up and dracut does the 'right' thing in the initrd but your 'init' will crap out
with this config since it won't be able to handle your swap.. the swapon call will try to access
your /dev/foo while should swapon the 'open' crypt swap partiton which need be the /dev/mapper/luks-uuid 

and ofc resume won't work there's wrong swap + no swap being enable in userspace

all that is true for all kind supported combos.

is really simple to reproduce and you don't need cala to see where your the bugs lies here.

create a swap luks partition and add the way cala doing now  and reboot your box.
watch in your log for swap failed on /dev/sdx .. can't find swap signature , check top etc 
and you should not have any swap.

run blkid and look for right /dev/mapper for your swap partition and add now like this 

resume=/dev/mapper/luks-the-uuid ...

and same in fstab , change UUID= with /dev/mapper/..

reboot your box and all should work.

**Comment by kkofler** [comment]
So you are seeing a different bug than what I am seeing on Fedora. I could not get `resume=UUID=…` to work at all. Resume did not even work for systems installed with Anaconda. And the Fedora developers don't care because resume is not supported at all with "Secure Boot" enabled and they don't care about setups with that misfeature disabled.

**Comment by abucodonosor** [comment]
ehh while 'secure boot' is  a misfeature ifself .. 

@kkofler 

do you have a livecd with fedora and calamaers as installer ? if so please give me a link and I can try to
figure what is wrong there.

**Comment by philmmanjaro** [comment]
@abucodonosor: you find his work [here](https://sourceforge.net/projects/kannolo/).

**Comment by abucodonosor** [comment]
@philmmanjaro 

thx , I have a look

**Comment by abucodonosor** [comment]
@kkofler 

I've tested your ISO in virtualbox and suspend to disk works just fine as is
Also  LUKS install with suspend to disk works fine 

But while at it notice when doing manual prtitioning ( since you have swap off by default ) 
calamares writes right swap LUKS uuid in ftsab at least..

So one more bug to look at in auto layout of calamares.

**Comment by kkofler** [comment]
Swap is supposed to be on by default, assuming your (virtualized?) disk is large enough.

**Comment by abucodonosor** [comment]
@kkofler 

I let it default 8GB mabye to small yes

**Comment by adriaandegroot** [comment]
Is this the same as #772? This issue also mentions GRUB, but I can't see what is expected behavior / expected file contents and observed behavior. This is probably my lack of experience with Linux encrypted systems, so please spell things out fully.

**Comment by abucodonosor** [comment]
@adriaandegroot 

#772 seems to be about rootfs on debian systems.. I never got calamares to have
wrong rootfs entry at all.

I can test these things with dracut+LUKS+systemd or mkinitcpio+LUKS+systemd and
the swap issue is still around.

**Comment by adriaandegroot** [comment]
I'm going to need a lot more information to be able to do anything with this; mostly because I'm not familiar with all of the configurations that there are. I don't know what dracut is supposed to do; it's not on any of the distro's I ever use.

 - What distro as base?
 - What settings exactly in the calamares partitioning page?
 - What are the file contents you get in the installed system?
 - What are the file contents you expect / need in the installed system?

@abucodonosor the patch you wrote looks suspect to me; if you were to do a pull request I would add review comments. Basically `dracut_bin == 0` means "dracut is installed" so it looks like some of the logic is backwards. You seem to be setting resume= to the luks mapper name of the root partition .. is that intentional?

**Comment by adriaandegroot** [comment]
I think I can reproduce this with Manjaro 17.04 (see #606), but that doesn't help much in debugging and fixing the underlying issue.

**Comment by philmmanjaro** [comment]
You mean with Manjaro **v17.0.3** which we just had released. I tried to add all fixes as given. Since you released some more, I've to merge them first.

**Comment by adriaandegroot** [comment]
No, 17.0.1 it turns out .. whatever I had available.

**Comment by abucodonosor** [comment]
@adriaandegroot 

I got my own modules for fstab,grub,bootloader the time .. 

So @demmm pulled in the mkinitcpio hunks in his own modules for KaOS
V3n3RiX pulled in the swap fix in fstab for his gentoo based distro.

Yes I know dracut_bin == 0 mens installed .. and yeah this patch is very ugly

So when dracut installed and when swap and swap is NOT luks, bc later on we check for swap_outer_uuid for dracut which means there is LUKS.

I changed resume= bc root= using already luks names also swap does now with the patch
you just pulled in from V3n3RiX.

Setup is default .. get a HDD >=16GB in VirtualBox , let all default layout and use crypt.
